Ensure `AUTOINCREMENT` declaration is preserved when altering SQLite tables

adrianna-chang-shopify opened the pull request:
### Motivation / Background

Closes: https://github.com/rails/rails/issues/47400

Previously, when copying a table as a result of an `ALTER TABLE` statement, the `AUTOINCREMENT` declaration on the PK column was dictated by the absence of a `:default` option on that column. However, we were always passing `default` when copying the columns to the new table, regardless of whether the column originally had a default. This led to the `AUTOINCREMENT` declaration being lost.

### Detail

This commit fixes the issue by tracking a column's `auto_increment` value, and only passing the `default` option in `#copy_table` when the column does not have the `auto_increment` option set.

I've added a test that ensure that both storing `auto_increment` on `SQLite3::Column` works as intended, and that the `AUTOINCREMENT` declaration is preserved on the table itself.

### Additional information

<!-- Provide additional information such as benchmarks, reference to other repositories or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

adrianna-chang-shopify committed:
Ensure auto_increment declaration is preserved when copying SQLite tables

Fixes #47400.
Previously, when copying a table as a result of an ALTER TABLE
statement, the AUTOINCREMENT declaration on the PK column was dictated
by the absence of a `:default` option on that column. However,
we were always passing `default` when copying the columns to the new table,
regardless of whether the column originally had a default. This led to
the AUTOINCREMENT declaration being lost.

This commit fixes the issue by tracking a column's auto_increment value,
and only passing the `default` option in `#copy_table` when the column
does not have the `auto_increment` option set.

Co-authored-by: fatkodima <fatkodima123@gmail.com>

yahonda commented:
These CI errors need addressed.

https://buildkite.com/rails/rails/builds/93919#01866019-18d4-412a-b74d-bbbd6b091c61/1109-1255
https://buildkite.com/rails/rails/builds/93919#01866019-18d4-412a-b74d-bbbd6b091c61/1109-1301

adrianna-chang-shopify committed:
Require SQLite3 column so that we can attempt to load the previous schema cache

Previously, the schema cache dump contained ActiveRecord::ConnectionAdapters::Column objects,
and requiring the app env file and initializing the application would load
the appropriate Column constant. Now, however, SQLite3 uses its own Column
class, so the schema cache dump contains ActiveRecord::ConnectionAdapters::SQLite3::Column
objects. We don't load the SQLite3 adapter, since we switch to use mysql2 in order
to simulate an unhealthy db, and consequently loading SQLite3::Column fails.

As a workaround, I've opted to simply require the SQLite3 column class.

adrianna-chang-shopify commented:
@yahonda I've fixed the failing railties tests in a separate commit because the failure was a bit tricky. Let me know what you think :)
